### PR TITLE
Checkout dir should be available as artifact from bwc projects

### PR DESCRIPTION
**Pull request by breskeby**
Related to #62985 as a step forward on improving resolving test resources
from older elasticsearch branches / versions

### CODE REVIEW

**Comment by rjernst** [comment]
I don't think we should be making the entire checkout dir available. In general, the bwc checkouts are only or testing bwc against snapshot versions. But we still need to test against released versions as well, where we don't have a checkout. By requiring that anything we rely on from the bwc project is an artifact, we ensure we can also test against that same artifact for released versions. The switch is just where we consume it from in the build (pointing to a bwc project artifact or a released maven artifact), but the test itself doesn't have to be written any differently. If we were to consume the entire checkout, a test could only be written against snapshot versions.

**Comment by breskeby** [comment]
@rjernst I agree on your concerns. We discussed this briefly with @jakelandis 
My concrete suggestion to solve this in a transparent way for the test author would be to 

1. Add a test fixture project that basically provides an artifact (directory / zip) that contains all test compatibility resources
2. Add a bwcTask similar to the `jdbc` bwc artifact to create that fixture artifact created in `1.` for bwc tests against these unreleased bwc projects and change the compatibility test plugin to resolve compatibility test resources from the new introduces unreleased versions from bwc subproject in `distribution:bwc...`.
4. Change our release process to publish versioned compatibility test resources in a zip as part of the release process
5. Add the functionality to the compatibility plugin that allows resolving compatibility test resources for released versions from external sources (e.g artefacts) as we do for released es versions in our existing test cluster setup. 

To not break things across branches we need to break this up in separate PRs I think

**Comment by jakelandis** [comment]
> Add a test fixture project that basically provides an artifact (directory / zip) that contains all test compatibility resources

Actually the zip file need to contain all `current` test resources which will be published at release time. That is different from what we currently publish since it needs to include the api/test from all modules/plugins, and x-pack:plugin and x-pack:plugin:*. That is what needs to be published on an ongoing basis. [1] We can manually curate those for prior branches as a one-off event. 

For this PR I need a single resolvable artifact that has all of the api/test modules/plugins, and x-pack:plugin and x-pack:plugin:* sourced from the bwc:minor branch (so we can test against current unreleased version). I don't have a preference if that artifact is whole cloned source tree, or a curated subset of that checkout with only apt/tests, or a custom built zip from the api/test. 

I think we can separate the need for a published zip (future PR?) and the need for a resolvable artifact sourced from bwc:minor (this PR?). It would be nice if the syntax between testing a released version and bwc:minor were the same but it isn't too bad for the plugin to if/else code against released versions and un-released version. 


[1] We actually do have a need to publish the compatible set of tests via the zip .. but exactly the way you describe, and we not really discussed that prior.... I will raise a discussion point outside of this PR.

**Comment by breskeby** [comment]
@elasticmachine test this please

**Comment by breskeby** [comment]
@jakelandis so as a first iteration I think providing the checkoutdir as artefact is the right approach then to have coverage for our bwc distributions. the zip later on could provide just provide the same structure for as the repo structure. We can revisit the structure later then if it makes sense. 

This way 

1. the tests wouldn't need to differentiate between the bwc checkout resources and zipped resources
2. we would not need an extra copy operation to copy the test resources into a different structure than how its spread across the repo

File: buildSrc/src/integTest/groovy/org/elasticsearch/gradle/internal/InternalBwcGitPluginFuncTest.groovy [hunk 1]
```diff
@@ -19,33 +19,60 @@
...
         def result = gradleRunner("createClone", '--stacktrace').build()
         then:
         result.task(":createClone").outcome == TaskOutcome.SUCCESS
-        file("build/checkout/build.gradle").exists()
-        file("build/checkout/settings.gradle").exists()
+        file("cloned/build/checkout/build.gradle").exists()
+        file("cloned/build/checkout/settings.gradle").exists()
+    }
+
+    def "can resolve checkout folder as project artifact"() {
```
Comment (breskeby) [inline buildSrc/src/integTest/groovy/org/elasticsearch/gradle/internal/InternalBwcGitPluginFuncTest.groovy:50]: @jakelandis this functional tests basically shows an example on how to resolve the checkoutDir as artifact. you can declare a dependency in the es build by declaring 

```
dependencies {
    myConfig project(path:':distribution:bwc:minor', configuration:'checkoutDir')
}

```

**Comment by mark-vieira** [comment]
I'm good with moving forward with handing over the checkout dir. It's still handled as a proper project dependencies and we seem to have a decent plan for iterating towards a better solution.

**Pull request merged**